From 8609f82a2c1fe5a32df189480060cf40921c8ff9 Mon Sep 17 00:00:00 2001
From: Brent Cook <busterb@gmail.com>
Date: Mon, 12 Jan 2015 06:18:31 -0600
Subject: [PATCH 09/12] initialize setproctitle where needed

We need to save a copy of argv and __progname to avoid setproctitle
clobbering them.
---
 src/usr.sbin/ntpd/ntpd.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/src/usr.sbin/ntpd/ntpd.c b/src/usr.sbin/ntpd/ntpd.c
index c7935bf..44caa80 100644
--- a/src/usr.sbin/ntpd/ntpd.c
+++ b/src/usr.sbin/ntpd/ntpd.c
@@ -112,6 +112,13 @@ usage(void)
 #define POLL_MAX		8
 #define PFD_PIPE		0
 
+/* Saves a copy of argv for setproctitle emulation */
+#ifndef HAVE_SETPROCTITLE
+static char **saved_argv;
+#endif
+
+char *get_progname(char *argv0);
+
 int
 main(int argc, char *argv[])
 {
@@ -135,6 +142,19 @@ main(int argc, char *argv[])
 
 	log_init(1);		/* log to stderr until daemonized */
 
+	__progname = get_progname(argv[0]);
+
+#ifndef HAVE_SETPROCTITLE
+	int i;
+	/* Prepare for later setproctitle emulation */
+	saved_argv = calloc(argc + 1, sizeof(*saved_argv));
+	for (i = 0; i < argc; i++)
+		saved_argv[i] = strdup(argv[i]);
+	saved_argv[i] = NULL;
+	compat_init_setproctitle(argc, argv);
+	argv = saved_argv;
+#endif
+
 	while ((ch = getopt(argc, argv, "df:np:sSv")) != -1) {
 		switch (ch) {
 		case 'd':
-- 
2.0.4

From 8b01bffe9f2def35b180118f9f41e47c4fc0a9cc Mon Sep 17 00:00:00 2001
From: "Paul B. Henson" <henson@acm.org>
Date: Sun, 5 Jul 2015 14:06:36 -0700
Subject: [PATCH 09/12 part 2] Move __progname initialization to before first use.

---
 src/usr.sbin/ntpd/ntpd.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/usr.sbin/ntpd/ntpd.c b/src/usr.sbin/ntpd/ntpd.c
index 44caa80..310e808 100644
--- a/src/usr.sbin/ntpd/ntpd.c
+++ b/src/usr.sbin/ntpd/ntpd.c
@@ -131,6 +131,8 @@ main(int argc, char *argv[])
 	struct passwd		*pw;
 	extern char		*__progname;
 
+	__progname = get_progname(argv[0]);
+
 	if (strcmp(__progname, "ntpctl") == 0) {
 		ctl_main(argc, argv);
 		/* NOTREACHED */
@@ -142,8 +144,6 @@ main(int argc, char *argv[])
 
 	log_init(1);		/* log to stderr until daemonized */
 
-	__progname = get_progname(argv[0]);
-
 #ifndef HAVE_SETPROCTITLE
 	int i;
 	/* Prepare for later setproctitle emulation */
-- 
2.0.4

